src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/ECDsaOpenSsl.cs

```diff
@@ -107,28 +111,63 @@ public override int KeySize
         {
             set
             {
-                if (KeySize == value)
+                if (_tempKeySize == 0)
                 {
-                    return;
-                }
+                    // if _tempKeySize is set, the key was already created or freed and should not be freed again
+                    if (KeySize == value)
+                    {
+                        return;
+                    }
+                    FreeKey();
 
-                FreeKey();
-                base.KeySize = value;
-                _key = new Lazy<SafeEcKeyHandle>(GenerateKey);
+                    base.KeySize = value;
+                    _key = new Lazy<SafeEcKeyHandle>(GenerateKey);
+                }
+                else
+                {
+                    base.KeySize = value;
+                }
             }
         }
 
+        private void SetKey(SafeEcKeyHandle key)
+        {
+            FreeKey();
+
+            _tempKeySize = Interop.Crypto.EcKeyGetSize(key);
+            KeySize = _tempKeySize;
+            _tempKeySize = 0;
+
+            _key = new Lazy<SafeEcKeyHandle>(() => key);
+        }
+
+        private static readonly KeySizes[] s_legalKeySizes =
+            new KeySizes[]
+            {
+                // todo: just keep original four keysizes? sync with cng code.
+              
+                // All values are in bits.
+                new KeySizes(minSize: 160, maxSize: 224, skipSize: 32), // 160, 192, 224
+                new KeySizes(minSize: 239, maxSize: 239, skipSize: 0),  // 239
+                new KeySizes(minSize: 255, maxSize: 255, skipSize: 0),  // 255
+                new KeySizes(minSize: 256, maxSize: 512, skipSize: 128),// 256, 384, 512
+                new KeySizes(minSize: 320, maxSize: 320, skipSize: 0),  // 320
+                new KeySizes(minSize: 521, maxSize: 521, skipSize: 0),  // 521
+            };
         public override KeySizes[] LegalKeySizes
         {
             get
             {
-                KeySizes[] legalKeySizes = new KeySizes[s_supportedAlgorithms.Length];
-                for (int i = 0; i < s_supportedAlgorithms.Length; i++)
+                KeySizes[] keySizes = (KeySizes[])s_legalKeySizes.Clone();
+
+                // if a curve dynamically add the keysize
+                if (_tempKeySize != 0)
```

Comment:
It looks like this is trying to do the same thing as `_skipKeySizeCheck` in RSAOpenSsl (https://github.com/dotnet/corefx/blob/master/src/Common/src/System/Security/Cryptography/RSAOpenSsl.cs#L88-L94); it would be good to have them operating in the same way.